First version of magic proxy service

hannahbast opened the pull request:
Implement a generic proxy service that sends arbitrary `sparql-results+json` input to a remote endpoint, receives `sparql-results+json` output from that endpoint, and joins this with the enclosing graph pattern. As usual for QLever's magic services, the configuration is provided via `_:config ql:proxy:... ...` triples. The input is provided via `ql:proxy-<input> ?var?`, where `?var` is a variable from the enclosing graph pattern, and `input` is the name of the variable sent to the endpoint. Additionally, URL parameters can be specified via `... qlproxy:param-<key> <value>` triples; they will be added to the endpoint request. The output is provided via `ql:proxy-<output> ?var?`, where `output` is the name of the variable sent back by the endpoint, and `?var` is the variable to which the respective results are bound in the query. The input sent to the endpoint is augmented by the variable specified by `ql:proxy:output-row`, which provides the row index for each binding. The endpoint must provide a binding for that same variable for each solution it returns. The result of the service request is joined with the result of the enclosing graph pattern on this variable. Here is an example query. A Python script `qlproxy_qgrams.py` that implements the endpoint for this example, is part of this PR.
```
PREFIX qlproxy: <https://qlever.cs.uni-freiburg.de/qlproxy/>
SELECT ?row ?word ?qgram WHERE {
  VALUES ?word { "doof" "bloed" }
  SERVICE qlproxy: {
    _:config qlproxy:endpoint <https://qlever.dev/api/qgram-service> ;
             qlproxy:input-word ?word ;
             qlproxy:output-qgram ?qgram ;
             qlproxy:output-row ?row ;
             qlproxy:param-q "3" ;
             qlproxy:param-padding "left" .
  }
}
```

Commit:
First version of magic proxy service

Includes `qlproxy_server.py` for end-to-end testing. This should
eventually be removed again. For example, when that server is started
with `python3 qlproxy_server.py --port 8888 verbose`, the the following
should work in QLever:

```sparql
PREFIX qlproxy: <https://qlever.cs.uni-freiburg.de/qlproxy/>
SELECT ?result WHERE {
  VALUES (?x1 ?x2) { (1 3) (2 2) (3 1) }
  SERVICE qlproxy: {
    _:config qlproxy:endpoint <https://qlever.dev/api/proxy-test> ;
             qlproxy:payload_first ?x1 ;
             qlproxy:payload_second ?x2 ;
             qlproxy:result_res ?result ;
             qlproxy:param_op "add" .
  }
}
```

Commit:
Forgot two files in the last commit

Commit:
Merge remote-tracking branch 'origin/master' into qlproxy-service

Commit:
Rename `ProxyOperation` to `Proxy` + fix C++17 runs

patrickbr reviewed (commented):
I tested this this morning and it works very well as long as there is a single proxy service. If you try to combine multiple proxy services, it is quite easy to generate cryptic assertion failures:

```
PREFIX qlproxy: <https://qlever.cs.uni-freiburg.de/qlproxy/>
SELECT ?result2 WHERE {
  VALUES (?x1 ?x2) { (1 3) (2 2) (3 1) }
  SERVICE qlproxy: {
    _:config qlproxy:endpoint <https://qlever.dev/api/proxy-test> ;
             qlproxy:payload_first ?x1 ;
             qlproxy:payload_second ?x2 ;
             qlproxy:result_res ?result ;
             qlproxy:param_op "add" .
  }
  SERVICE qlproxy: {
    _:config qlproxy:endpoint <https://qlever.dev/api/proxy-test> ;
             qlproxy:payload_first ?x1 ;
             qlproxy:payload_second ?x2 ;
             qlproxy:result_res ?result2 ;
             qlproxy:param_op "add" .
  }
}
```
results in an `ERROR: Assertion 'res > 0' failed.` in `MultiColumnJoin.cpp`. I guess that query doesn't make any sense because the variables `?x1` and `?x2` are not "passed through" by the proxy, but the error message should be better.

For the following query, I would naively expect a result of

```
8
8
8
```

but I am getting `ERROR: Assertion 'variablesAreDisjoint' failed.`:

```
PREFIX qlproxy: <https://qlever.cs.uni-freiburg.de/qlproxy/>
SELECT ?result2 WHERE {
  VALUES (?x1 ?x2) { (1 3) (2 2) (3 1) }
  SERVICE qlproxy: {
    _:config qlproxy:endpoint <https://qlever.dev/api/proxy-test> ;
             qlproxy:payload_first ?x1 ;
             qlproxy:payload_second ?x2 ;
             qlproxy:result_res ?result ;
             qlproxy:param_op "add" .
  }
  SERVICE qlproxy: {
    _:config qlproxy:endpoint <https://qlever.dev/api/proxy-test> ;
             qlproxy:payload_first ?result ;
             qlproxy:payload_second ?result ;
             qlproxy:result_res ?result2 ;
             qlproxy:param_op "add" .
  }
}
```

However, I am not sure whether queries which include multiple `SERVICE` proxy calls like that even make any sense.

EDIT: in the query above, the following works:

```
PREFIX qlproxy: <https://qlever.cs.uni-freiburg.de/qlproxy/>
SELECT ?result2 WHERE {
  {
    SELECT * WHERE {
      VALUES (?x1 ?x2) { (1 3) (2 2) (3 1) }
      SERVICE qlproxy: {
        _:config qlproxy:endpoint <https://qlever.dev/api/proxy-test> ;
                 qlproxy:payload_first ?x1 ;
                 qlproxy:payload_second ?x2 ;
                 qlproxy:result_res ?result ;
                 qlproxy:param_op "add" .
      }
    }
  }
  SERVICE qlproxy: {
    _:config qlproxy:endpoint <https://qlever.dev/api/proxy-test> ;
             qlproxy:payload_first ?result ;
             qlproxy:payload_second ?result ;
             qlproxy:result_res ?result2 ;
             qlproxy:param_op "add" .
  }
}
```

Commit:
Merge remote-tracking branch 'origin/master' into qlproxy-service

Commit:
Add row column and improve syntax

Commit:
Add qgram-proxy service, as a more interesting example

Commit:
Add cache key and consider when `cache-service-results` is on

Commit:
Merge remote-tracking branch 'origin/master' into qlproxy-service

Commit:
Merge remote-tracking branch 'origin/master' into qlproxy-service

Commit:
Merge remote-tracking branch 'origin/master' into qlproxy-service

Commit:
Merge remote-tracking branch 'origin/master' into qlproxy-service

Commit:
Deep-clone of the children of the proxy operation

The previous implementation (which was written as part of this brach)
was not correct: only a shared pointer to the children was created.

hannahbast committed:
Merge remote-tracking branch 'origin/master' into qlproxy-service

hannahbast committed:
Fix codespell problen
